lib/ruby_lsp/tapioca/server_addon.rb

```diff
@@ -14,6 +14,33 @@ def name
 
       def execute(request, params)
         case request
+        when "reload_workspace_compilers"
+          with_notification_wrapper("reload_workspace_compilers", "Reloading DSL compilers") do
+            compiler_paths = Dir.glob(
+              "#{params[:workspace_path]}/**/tapioca/**/compilers/**/*.rb",
+              File::FNM_PATHNAME | File::Constants::FNM_EXTGLOB,
+            )
+
+            # Remove all loaded compilers that are inside the workspace
+            ::Tapioca::Dsl::Compiler.descendants.each do |compiler|
+              name = compiler.name
+              next unless name && compiler_paths.include?(Module.const_source_location(name).first)
+
+              *parts, unqualified_name = name.split("::")
+
+              if parts.empty?
+                Object.send(:remove_const, unqualified_name)
+              else
+                parts.join("::").safe_constantize.send(:remove_const, unqualified_name)
+              end
+            end
+
+            # Remove from $LOADED_FEATURES each workspace compiler file and then re-required it to reload
+            compiler_paths.each do |path|
+              $LOADED_FEATURES.delete(path)
+              require File.expand_path(path)
+            end
```

Comment:
I would very much like to see this encapsulated in the DSL pipeline logic that actually does the compiler loading and here is why:

The DSL pipeline already knows the exact paths and names of the compilers that it has loaded, and it can keep track of those for this purpose. When we ask the pipeline to reload its compilers, we could give it an (optional) list of paths that have changed, and it can decide if any of the compilers will be loaded or not. And the LSP logic will be much clearer.

Reply:
I agree with you, but we once again hit the problem that we lose the whole pipeline state due to forking.

In the latest commit, I moved the logic to `Loaders::Dsl`, since we can easily hold a reference to a loader and use it repeatedly. Let me know what you think or you feel like I'm missing something.

Reply:
Yes, sorry, that was exactly what I had in mind, thank you. I misremembered it as the pipeline loading the compilers, but it wasn't, it was the loader, of course.

This looks great, thanks for the refactor. IMO, this looks much nicer and is more encapsulated.